### DISCUSSION TITLE
Symmetric boundary condition

### DISCUSSION THREAD

**Discussion by simbilod** [kinnala/scikit-fem#1112] [Q&A]
Hello kinnala and other users/developers, great library!

Does there already exist a way to define symmetric (mirror) boundary conditions (for example as described [here](https://www.simscale.com/docs/simulation-setup/boundary-conditions/symmetry/))?

If not, is there something can we easily modify to enable this?

**Comment by kinnala** [comment]
It is possible to impose such symmetry conditions. Usually symmetry conditions are implemented in terms of other basic boundary condition types (Neumann or Dirichlet). The practical implementation depends, to some extent, on which governing equation you are interested in. Is it linear elasticity as in your link?

**Reply by simbilod** [reply]
Thanks for the quick reply! 

I am actually using scikit-fem for electromagnetic eigenvalue solving. So the field I am solving for is vectorial, and according to that link it seems I need to set the derivative of all components to 0, and the values of the normal components to 0? 

Below I try different boundaries (fully Dirichlet for full, and Dirichlet on 3 sides and Neumann on the left for half), but I think I am not tackling the components separately. 

You can set full to True for the full geometry:

![image](https://github.com/kinnala/scikit-fem/assets/46427609/63a3629e-5a1e-4205-9ccb-c89340a9fff8)
![image](https://github.com/kinnala/scikit-fem/assets/46427609/c9390b32-1c09-4c0e-9f6f-e21a13e6b24d)

and False for the half geometry, which should return the same result under mirror symmetry for the lowest-order mode being solved for.

![image](https://github.com/kinnala/scikit-fem/assets/46427609/80c3be87-fe70-470f-b1d3-1e85026f1006)
![image](https://github.com/kinnala/scikit-fem/assets/46427609/7e91fe77-470a-45eb-86c4-0624e319cb19)


Code:

```
from dataclasses import dataclass
from functools import cached_property
from typing import Callable, List, Literal, Optional, Tuple
from warnings import warn

import matplotlib.pyplot as plt
import numpy as np
import scipy.constants
import scipy.sparse.linalg
from matplotlib.axes import Axes
from matplotlib.figure import Figure
from numpy.typing import NDArray
from scipy.constants import epsilon_0, speed_of_light
from skfem import (
    Basis,
    BilinearForm,
    ElementDG,
    ElementTriN1,
    ElementTriN2,
    ElementTriP0,
    ElementTriP1,
    ElementTriP2,
    ElementVector,
    Functional,
    InteriorFacetBasis,
    LinearForm,
    Mesh,
    condense,
    solve,
)
from skfem.helpers import cross, curl, dot, grad, inner
from skfem.utils import solver_eigen_scipy
from mpl_toolkits.axes_grid1 import make_axes_locatable

"""
GEOMETRY + BOUNDARIES
"""

full = True

if full:
    mesh = Mesh.load("full_mesh.msh")
    xmin, xmax, ymin, ymax = np.min(mesh.doflocs[0,:]), np.max(mesh.doflocs[0,:]), np.min(mesh.doflocs[1,:]), np.max(mesh.doflocs[1,:])
    boundaries = {"bottom": lambda xi: xi[1] == ymin,
                                "top": lambda xi: xi[0] == ymax,
                                "right": lambda xi: xi[1] == xmax,
                                "left": lambda xi: xi[0] == xmin,
                                }
else:
    mesh = Mesh.load("half_mesh.msh")
    xmin, xmax, ymin, ymax = np.min(mesh.doflocs[0,:]), np.max(mesh.doflocs[0,:]), np.min(mesh.doflocs[1,:]), np.max(mesh.doflocs[1,:])
    boundaries = {"bottom": lambda xi: xi[1] == ymin,
                                "top": lambda xi: xi[0] == ymax,
                                "right": lambda xi: xi[1] == xmax,
                                }

mesh = mesh.with_boundaries(boundaries)


# Domain
basis_epsilon_r = Basis(mesh, ElementTriP0(), intorder=4)
epsilon_r = basis_epsilon_r.zeros(dtype=float)

for subdomain, n in {
    "silicon_input": 3.45,
    "silicon_detector": 3.45,
    "germanium": 4.6530,
    "clad": 1.44,
}.items():
    epsilon_r[basis_epsilon_r.get_dofs(elements=subdomain)] = n**2

"""
SOLVING
"""
# Parameters
wavelength = 1.55
order = 1
k0 = 2 * np.pi / wavelength
mu_r = 1
num_modes=1

if order == 1:
    element = ElementTriN1() * ElementTriP1()
elif order == 2:
    element = ElementTriN2() * ElementTriP2()
else:
    raise AssertionError("Only order 1 and 2 implemented by now.")

basis = basis_epsilon_r.with_element(element)
basis_epsilon_r = basis.with_element(basis_epsilon_r.elem)  # adjust quadrature

@BilinearForm(dtype=epsilon_r.dtype)
def aform(e_t, e_z, v_t, v_z, w):
    epsilon = w.epsilon

    return (
        1 / mu_r * curl(e_t) * curl(v_t) / k0**2
        - epsilon * dot(e_t, v_t)
        + 1 / mu_r * dot(grad(e_z), v_t)
        + epsilon * inner(e_t, grad(v_z))
        - epsilon * e_z * v_z * k0**2
    )

@BilinearForm(dtype=epsilon_r.dtype)
def bform(e_t, e_z, v_t, v_z, w):
    return -1 / mu_r * dot(e_t, v_t) / k0**2

A = aform.assemble(basis, epsilon=basis_epsilon_r.interpolate(epsilon_r))
B = bform.assemble(basis, epsilon=basis_epsilon_r.interpolate(epsilon_r))

sigma = sigma = k0**2 * np.max(epsilon_r) * 1.1

lams, xs = solve(
    *condense(
        -A,
        -B,
        D=basis.get_dofs(set(boundaries.keys())),
        x=basis.zeros(dtype=float),
    ),
    solver=solver_eigen_scipy(k=num_modes, sigma=sigma),
)

# Plotting
mode = xs[:, 0]
colorbar = "same"
(et, et_basis), (ez, ez_basis) = basis.split(mode)

plot_basis = et_basis.with_element(ElementVector(ElementDG(ElementTriP1())))
et_xy = plot_basis.project(et_basis.interpolate(et))
(et_x, et_x_basis), (et_y, et_y_basis) = plot_basis.split(et_xy)

rc = (3, 1)
fig, axs = plt.subplots(*rc, subplot_kw=dict(aspect=1))
for ax in axs:
    basis.mesh.draw(ax=ax, boundaries_only=True)
    for subdomain in basis.mesh.subdomains.keys() - {"gmsh:bounding_entities"}:
        basis.mesh.restrict(subdomain).draw(ax=ax, boundaries_only=True)

for ax, component in zip(axs, "xyz"):
    ax.set_title(f"$field_{component}$")

maxabs = max(np.max(np.abs(data.value)) for data in basis.interpolate(mode))
vmin = -maxabs if colorbar == "same" else None
vmax = maxabs if colorbar == "same" else None

et_x_basis.plot(et_x, shading="gouraud", ax=axs[0], vmin=vmin, vmax=vmax)
et_y_basis.plot(et_y, shading="gouraud", ax=axs[1], vmin=vmin, vmax=vmax)
ez_basis.plot(ez, shading="gouraud", ax=axs[2], vmin=vmin, vmax=vmax)

if colorbar:
    if colorbar == "same":
        plt.colorbar(axs[0].collections[-1], ax=axs.ravel().tolist())
    else:
        for ax in axs:
            divider = make_axes_locatable(ax)
            cax = divider.append_axes("right", size="5%", pad=0.05)
            plt.colorbar(ax.collections[-1], cax=cax)
        plt.tight_layout()

plt.show()
```

And here are the mesh files (they should be .msh files, but Github does not allow uploading that format):

[full_mesh.txt](https://github.com/kinnala/scikit-fem/files/14765320/full_mesh.txt)
[half_mesh.txt](https://github.com/kinnala/scikit-fem/files/14765321/half_mesh.txt)

**Reply by simbilod** [reply]
Also, anti-symmetric boundary conditions would be great to model as well, which I think we can get by switching out which components are pinned and which are free

**Reply by kinnala** [reply]
As the basis functions of Nedelec elements are vector-valued, it may not be possible to set specific components to zero. I tried to find information on how symmetry conditions are implemented for Nedelec elements but I was not able to find anything. I will do a more careful review at some point. I think the way it is done for $H^1$ conforming elements and $H(curl)$ conforming elements will be different.

**Reply by simbilod** [reply]
Sounds good! I'm not extremely familiar with the FEM literature, so if you have anything to recommend me to look into I can do that too

**Reply by kinnala** [reply]
I have a background in solid mechanics so I can intuitively understand how symmetry conditions work there. The point is to not allow any displacement in the normal direction with respect to the symmetry plane as otherwise there would be a gap in the material over the symmetry plane. The tangential direction should be traction free as there are not any external forces acting there. These will involve a dot product of the displacement and the normal vector and the dot product of the traction vector and the normal vector.

I have no intuition for electromagnetism, I studied just one course more than ten years ago, and it involved no numerics. According to this source, the symmetry conditions in electromagnetism may include the cross product of the field and the normal vector: https://onlinelibrary.wiley.com/doi/abs/10.1002/nme.6088 I am not sure how this can be physically interpreted but perhaps it makes more sense to you?

**Reply by HelgeGehring** [reply]
It seems to me that we just find the second eigenmode with the current boundary conditions, right?
If we just stick with the boundary conditions of "full" I get exactly the mode we see for full.

I somewhen thought quite a bit about this (DOFs of nedelec element = 0 at the interface => tangential vector at interface = 0, right?) for doing perflectly magnetic / electric conductors, but I didn't write it down yet :/

**Reply by kinnala** [reply]
Yes, Nedelec DOFs correspond to the tangential component of the vector field.
